### ISSUE TITLE
std::out_of_range thrown during int8 engine creation

### ISSUE THREAD

**Issue by WaiTsun-Yeung** [marcoslucianops/DeepStream-Yolo#79]
We were trying out the steps on the ReadMe file on a Jetson platform to confirm it works as expected. We tried it out with the provided yolov3-tiny configuration and weight files. It ran successfully following the steps under the section _Basic Usage._ However, after we moved on to try out the steps in _INT8 calibration, _ the cmake command didn't generate the model_b1_gpu0_int8.engine file as expected, which prevented us from trying to run the algorithm in int8.

**Comment by WaiTsun-Yeung** [comment]
On another note, the calib.table file also has not been generated from any of the above steps and it is not clear what should be done to generate them.

**Comment by WaiTsun-Yeung** [comment]
I added some debugging printouts at the Yolo.cpp file at the createEngine function:

      nvinfer1::ICudaEngine *Yolo::createEngine (nvinfer1::IBuilder* builder)
          {
          assert (builder);
      
          m_ConfigBlocks = parseConfigFile(m_ConfigFilePath);
          parseConfigBlocks();
          orderParams(&m_OutputMasks);
      
	      std::cout << "Create engine" << std::endl;
      
      
          if (m_NetworkMode == "INT8" && !fileExists(m_Int8CalibPath)) {
              assert(builder->platformHasFastInt8());
	      std::cout << "network mode int8 received" << std::endl;
       ......

 and after I typed in this command at the terminal:

`sudo CUDA_VER=10.2 OPENCV=1 make -C nvdsinfer_custom_impl_Yolo`

 it returned the following message

`make: Entering directory '/opt/nvidia/deepstream/deepstream-5.1/sources/DeepStream-Yolo/native/nvdsinfer_custom_impl_Yolo'
g++ -c -DOPENCV -o yolo.o -Wall -std=c++11 -shared -fPIC -Wno-error=deprecated-declarations -I/opt/nvidia/deepstream/deepstream-5.1/sources/includes -I/usr/local/cuda-10.2/include -I/usr/include/opencv4/opencv -I/usr/include/opencv4 yolo.cpp
yolo.cpp: In member function ‘nvinfer1::ICudaEngine* Yolo::createEngine(nvinfer1::IBuilder*)’:
yolo.cpp:100:34: warning: ‘virtual void nvinfer1::IBuilder::setInt8Mode(bool)’ is deprecated [-Wdeprecated-declarations]
         builder->setInt8Mode(true);
                                  ^
In file included from layers/convolutional_layer.h:12:0,
                 from yolo.h:29,
                 from yolo.cpp:26:
/usr/include/aarch64-linux-gnu/NvInfer.h:6963:33: note: declared here
     TRT_DEPRECATED virtual void setInt8Mode(bool mode) TRTNOEXCEPT = 0;
                                 ^~~~~~~~~~~
yolo.cpp:101:46: warning: ‘virtual void nvinfer1::IBuilder::setInt8Calibrator(nvinfer1::IInt8Calibrator*)’ is deprecated [-Wdeprecated-declarations]
         builder->setInt8Calibrator(calibrator);
                                              ^
In file included from layers/convolutional_layer.h:12:0,
                 from yolo.h:29,
                 from yolo.cpp:26:
/usr/include/aarch64-linux-gnu/NvInfer.h:6979:33: note: declared here
     TRT_DEPRECATED virtual void setInt8Calibrator(IInt8Calibrator* calibrator) TRTNOEXCEPT = 0;
                                 ^~~~~~~~~~~~~~~~~
yolo.cpp:111:68: warning: ‘virtual nvinfer1::INetworkDefinition* nvinfer1::IBuilder::createNetwork()’ is deprecated [-Wdeprecated-declarations]
     nvinfer1::INetworkDefinition *network = builder->createNetwork();
                                                                    ^
In file included from layers/convolutional_layer.h:12:0,
                 from yolo.h:29,
                 from yolo.cpp:26:
/usr/include/aarch64-linux-gnu/NvInfer.h:6799:58: note: declared here
     TRT_DEPRECATED virtual nvinfer1::INetworkDefinition* createNetwork() TRTNOEXCEPT = 0;
                                                          ^~~~~~~~~~~~~
yolo.cpp:123:71: warning: ‘virtual nvinfer1::ICudaEngine* nvinfer1::IBuilder::buildCudaEngine(nvinfer1::INetworkDefinition&)’ is deprecated [-Wdeprecated-declarations]
     nvinfer1::ICudaEngine * engine = builder->buildCudaEngine(*network);
                                                                       ^
In file included from layers/convolutional_layer.h:12:0,
                 from yolo.h:29,
                 from yolo.cpp:26:
/usr/include/aarch64-linux-gnu/NvInfer.h:6934:51: note: declared here
     TRT_DEPRECATED virtual nvinfer1::ICudaEngine* buildCudaEngine(
                                                   ^~~~~~~~~~~~~~~
g++ -o libnvdsinfer_custom_impl_Yolo.so  nvdsinfer_yolo_engine.o nvdsparsebbox_Yolo.o yoloPlugins.o layers/convolutional_layer.o layers/dropout_layer.o layers/shortcut_layer.o layers/route_layer.o layers/upsample_layer.o layers/maxpool_layer.o layers/activation_layer.o utils.o yolo.o yoloForward.o calibrator.o -shared -Wl,--start-group -lopencv_dnn -lopencv_gapi -lopencv_highgui -lopencv_ml -lopencv_objdetect -lopencv_photo -lopencv_stitching -lopencv_video -lopencv_calib3d -lopencv_features2d -lopencv_flann -lopencv_videoio -lopencv_imgcodecs -lopencv_imgproc -lopencv_core -lnvinfer_plugin -lnvinfer -lnvparsers -L/usr/local/cuda-10.2/lib64 -lcudart -lcublas -lstdc++fs -Wl,--end-group
make: Leaving directory '/opt/nvidia/deepstream/deepstream-5.1/sources/DeepStream-Yolo/native/nvdsinfer_custom_impl_Yolo'
`

**Comment by marcoslucianops** [comment]
Hi, the make command doesn't generate the calibration files. You need to run the deepstream-app command to generate after modifying the config_primary_infer.txt file (to enable the INT8 mode).

**Comment by WaiTsun-Yeung** [comment]
Do you mean changing the engine file path and network mode? But the _model_b1_gpu0_int8.engine_ file is missing. If I tried to run `deepstream-app -c deepstream_app_config.txt` with this _config_primary_infer.txt_ file: 

      [property]
      gpu-id=0
      net-scale-factor=0.0039215697906911373
      model-color-format=0
      custom-network-config=yolov3-tiny.cfg
      model-file=yolov3-tiny.weights
      model-engine-file=model_b1_gpu0_int8.engine
      int8-calib-file=calib.table
      labelfile-path=labels.txt
      batch-size=1
      network-mode=1
      num-detected-classes=80
      interval=0
      gie-unique-id=1
      process-mode=1
      network-type=0
      cluster-mode=4
      maintain-aspect-ratio=0
      parse-bbox-func-name=NvDsInferParseYolo
      custom-lib-path=nvdsinfer_custom_impl_Yolo/libnvdsinfer_custom_impl_Yolo.so
      engine-create-func-name=NvDsInferYoloCudaEngineGet
      
      [class-attrs-all]
      pre-cluster-threshold=0.25

the terminal returned the following printout:

      Using winsys: x11 
      ERROR: Deserialize engine failed because file path: /opt/nvidia/deepstream/deepstream-5.1/sources/DeepStream-Yolo/native/model_b1_gpu0_int8.engine open error
      0:00:04.591495241 12018     0x24a68c10 WARN                 nvinfer gstnvinfer.cpp:616:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::deserializeEngineAndBackend() <nvdsinfer_context_impl.cpp:1691> [UID = 1]: deserialize engine from file :/opt/nvidia/deepstream/deepstream-5.1/sources/DeepStream-Yolo/native/model_b1_gpu0_int8.engine failed
      0:00:04.591693804 12018     0x24a68c10 WARN                 nvinfer gstnvinfer.cpp:616:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:1798> [UID = 1]: deserialize backend context from engine from file :/opt/nvidia/deepstream/deepstream-5.1/sources/DeepStream-Yolo/native/model_b1_gpu0_int8.engine failed, try rebuild
      0:00:04.591738540 12018     0x24a68c10 INFO                 nvinfer gstnvinfer.cpp:619:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:1716> [UID = 1]: Trying to create engine from model files
      WARNING: INT8 calibration file not specified/accessible. INT8 calibration can be done through setDynamicRange API in 'NvDsInferCreateNetwork' implementation
      Create engine
      File does not exist: /opt/nvidia/deepstream/deepstream-5.1/sources/DeepStream-Yolo/native/calib.table
      network mode int8 received
      
      Loading pre-trained weights
      Loading weights of yolov3-tiny complete
      Total weights read: 8858734
      
      Loading pre-trained weights
      Loading weights of yolov3-tiny complete
      Total weights read: 8858734
      Building YOLO network
            layer                        input               outup          weightPtr
      (0)   conv_leaky                 3 x 416 x 416      16 x 416 x 416    496    
      (1)   maxpool                   16 x 416 x 416      16 x 208 x 208    496    
      (2)   conv_leaky                16 x 208 x 208      32 x 208 x 208    5232   
      (3)   maxpool                   32 x 208 x 208      32 x 104 x 104    5232   
      (4)   conv_leaky                32 x 104 x 104      64 x 104 x 104    23920  
      (5)   maxpool                   64 x 104 x 104      64 x  52 x  52    23920  
      (6)   conv_leaky                64 x  52 x  52     128 x  52 x  52    98160  
      (7)   maxpool                  128 x  52 x  52     128 x  26 x  26    98160  
      (8)   conv_leaky               128 x  26 x  26     256 x  26 x  26    394096 
      (9)   maxpool                  256 x  26 x  26     256 x  13 x  13    394096 
      (10)  conv_leaky               256 x  13 x  13     512 x  13 x  13    1575792
      (11)  maxpool                  512 x  13 x  13     512 x  13 x  13    1575792
      (12)  conv_leaky               512 x  13 x  13    1024 x  13 x  13    6298480
      (13)  conv_leaky              1024 x  13 x  13     256 x  13 x  13    6561648
      (14)  conv_leaky               256 x  13 x  13     512 x  13 x  13    7743344
      (15)  conv_linear              512 x  13 x  13     255 x  13 x  13    7874159
      (16)  yolo                     255 x  13 x  13     255 x  13 x  13    7874159
      (17)  route                           -            256 x  13 x  13    7874159
      (18)  conv_leaky               256 x  13 x  13     128 x  13 x  13    7907439
      (19)  upsample                 128 x  13 x  13     128 x  26 x  26        -  
      (20)  route                           -            384 x  26 x  26    7907439
      (21)  conv_leaky               384 x  26 x  26     256 x  26 x  26    8793199
      (22)  conv_linear              256 x  26 x  26     255 x  26 x  26    8858734
      (23)  yolo                     255 x  26 x  26     255 x  26 x  26    8858734
      Output YOLO blob names: 
      yolo_17
      yolo_24
      Total number of YOLO layers: 46
      Building YOLO network complete
      Building the TensorRT Engine
      INFO: [TRT]: Detected 1 inputs and 2 output network tensors.
      INFO: [TRT]: Starting Calibration with batch size 1.
      INFO: [TRT]:   Post Processing Calibration data in 2.0833e-05 seconds.
      terminate called after throwing an instance of 'std::out_of_range'
        what():  _Map_base::at

**Comment by marcoslucianops** [comment]
do you have images in calibration folder, the calibration.txt file referring to the path of these images and the correct env variables?

**Comment by WaiTsun-Yeung** [comment]
Yes, I do. However, I am really new to this, and I am not sure how to convert calibration.txt to calib.table, and what do they have to do with the int8 engine file.

**Comment by WaiTsun-Yeung** [comment]
Now that I have reread the README.md file, I believe I have set the environment variables at the wrong places. Where should I include them in? I have appended them to the ~/.bashrc file.

**Comment by WaiTsun-Yeung** [comment]
I now have run `deepstream-app -c deepstream_app_config.txt` with this _config_primary_infer.txt_  file:

    [property]
    gpu-id=0
    net-scale-factor=0.0039215697906911373
    model-color-format=0
    custom-network-config=yolov3-tiny.cfg
    model-file=yolov3-tiny.weights
    # model-engine-file=model_b1_gpu0_fp32.engine
    # int8-calib-file=calib.table
    labelfile-path=labels.txt
    batch-size=1
    network-mode=1
    num-detected-classes=80
    interval=0
    gie-unique-id=1
    process-mode=1
    network-type=0
    cluster-mode=4
    maintain-aspect-ratio=0
    parse-bbox-func-name=NvDsInferParseYolo
    custom-lib-path=nvdsinfer_custom_impl_Yolo/libnvdsinfer_custom_impl_Yolo.so
    engine-create-func-name=NvDsInferYoloCudaEngineGet
    
    [class-attrs-all]
    pre-cluster-threshold=0.25

after exporting the correct env variables, the calibration.txt file is structured in this manner:

    /opt/nvidia/deepstream/deepstream-5.1/sources/DeepStream-Yolo/native/calibration/000000000285.jpg
    /opt/nvidia/deepstream/deepstream-5.1/sources/DeepStream-Yolo/native/calibration/000000000724.jpg
    /opt/nvidia/deepstream/deepstream-5.1/sources/DeepStream-Yolo/native/calibration/000000000776.jpg
    .....

and the the terminal returned the following printout:

    Using winsys: x11 
    0:00:00.366432900 23679   0x7f20002260 INFO                 nvinfer gstnvinfer.cpp:619:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() <nvdsinfer_context_impl.cpp:1716> [UID = 1]: Trying to create engine from model files
    WARNING: INT8 calibration file not specified/accessible. INT8 calibration can be done through setDynamicRange API in 'NvDsInferCreateNetwork' implementation
    Create engine
    File does not exist: 
    network mode int8 received
    
    Loading pre-trained weights
    Loading weights of yolov3-tiny complete
    Total weights read: 8858734
    
    Loading pre-trained weights
    Loading weights of yolov3-tiny complete
    Total weights read: 8858734
    Building YOLO network
          layer                        input               outup          weightPtr
    (0)   conv_leaky                 3 x 416 x 416      16 x 416 x 416    496    
    (1)   maxpool                   16 x 416 x 416      16 x 208 x 208    496    
    (2)   conv_leaky                16 x 208 x 208      32 x 208 x 208    5232   
    (3)   maxpool                   32 x 208 x 208      32 x 104 x 104    5232   
    (4)   conv_leaky                32 x 104 x 104      64 x 104 x 104    23920  
    (5)   maxpool                   64 x 104 x 104      64 x  52 x  52    23920  
    (6)   conv_leaky                64 x  52 x  52     128 x  52 x  52    98160  
    (7)   maxpool                  128 x  52 x  52     128 x  26 x  26    98160  
    (8)   conv_leaky               128 x  26 x  26     256 x  26 x  26    394096 
    (9)   maxpool                  256 x  26 x  26     256 x  13 x  13    394096 
    (10)  conv_leaky               256 x  13 x  13     512 x  13 x  13    1575792
    (11)  maxpool                  512 x  13 x  13     512 x  13 x  13    1575792
    (12)  conv_leaky               512 x  13 x  13    1024 x  13 x  13    6298480
    (13)  conv_leaky              1024 x  13 x  13     256 x  13 x  13    6561648
    (14)  conv_leaky               256 x  13 x  13     512 x  13 x  13    7743344
    (15)  conv_linear              512 x  13 x  13     255 x  13 x  13    7874159
    (16)  yolo                     255 x  13 x  13     255 x  13 x  13    7874159
    (17)  route                           -            256 x  13 x  13    7874159
    (18)  conv_leaky               256 x  13 x  13     128 x  13 x  13    7907439
    (19)  upsample                 128 x  13 x  13     128 x  26 x  26        -  
    (20)  route                           -            384 x  26 x  26    7907439
    (21)  conv_leaky               384 x  26 x  26     256 x  26 x  26    8793199
    (22)  conv_linear              256 x  26 x  26     255 x  26 x  26    8858734
    (23)  yolo                     255 x  26 x  26     255 x  26 x  26    8858734
    Output YOLO blob names: 
    yolo_17
    yolo_24
    Total number of YOLO layers: 46
    Building YOLO network complete
    Building the TensorRT Engine
    INFO: [TRT]: Detected 1 inputs and 2 output network tensors.
    INFO: [TRT]: Starting Calibration with batch size 1.
    INFO: [TRT]:   Post Processing Calibration data in 2.4097e-05 seconds.
    terminate called after throwing an instance of 'std::out_of_range'
      what():  _Map_base::at
    Aborted (core dumped)

**Comment by marcoslucianops** [comment]
It seems something wrong with your images, path in calibration.txt file or OpenCV installation.

**Comment by WaiTsun-Yeung** [comment]
I have run the script successfully. I realised it was installed in a root directory and needs root read and write permissions. Thank you so much for troubleshooting with me.
